docs/TensorLayout.md

```diff
@@ -0,0 +1,281 @@
+The struct contains the following fields:
+
+- `size_t alignment`
+	- Expected alignment of the current dimension.
+- `uint8_t order`
+	- Expected order of the current dimension.
+- `char name`
+	- Expected name of the current dimension.
+
+As an example, lets assume we want to create the array-of-structs for Glow's
+backend-agnostic `NHWC` layout, it should look as follows:
+
+```
+static TensorDimensionDescription dimsNHWC[] = {
+    {.alignment = 1, .order = 0, .name = 'N'},
+    {.alignment = 1, .order = 1, .name = 'H'},
+    {.alignment = 1, .order = 2, .name = 'W'},
+    {.alignment = 1, .order = 3, .name = 'C'},
+};
+```
+
+The `alignment` field has been set to `1` for all dimensions because
+we do not have an alignment requirement pre-lowering.
+
+Given the `N` dimension, for example, we set the `order` value to `0`
+indicating that it is the first dimension of tensor.
+
+To help indicate its purpose, number of samples, we give it the name `N`.
+
+Since some operations can take any layout, giving we use the `*` character
+to annotate a dimension as wild.
+
+As such, the description of a four dimensional tensor for a data parallel
+operation would look as follows:
+
+```
+static TensorDimensionDescription dims4D[] = {
+    {.alignment = 1, .order = 0, .name = '*'},
+    {.alignment = 1, .order = 1, .name = '*'},
+    {.alignment = 1, .order = 2, .name = '*'},
+    {.alignment = 1, .order = 3, .name = '*'},
+};
+```
+
+It is *highly* recommended to explicitly pass the expected layout for any operation
+that is not data parallel.
+It is not, however, a strict requirement: if it is not provided the
+operation will be constructed with `*` layout.
+
+This is done for connivence and backward compatibility's sake. If the layouts
+are dropped the default verifier will assume the operation will produce an
+output is in `NHWC` format.
+
+This does mean the verifier *might* fail if some operations have been annotated
+with `NCHW` while others are left as a wildcard. In a release build, if
+verifications are disabled, and assuming the graph is valid even if not
+correctly annotated, this will not cause a correctness issue on an
+Interpreter backend. This is not guaranteed on backends that use this (invalid)
+information for their backend-specific lowering and/or optimization.
+
+Another thing to note is that most optimization passes in Glow do not take
+a backend as an input, they might, however, be called post lowering. Layout
+verification will catch any (layout-requiring) transformation that breaks
+correctness due to implicitly assuming a canonical layout post-lowering.
+If a transformation requires Glow's canonical layout, it is its responsibility
+to verify that assumption.
+
+## Layout Requirements Interface
+
+Backends in Glow *may* derive from [base class `TensorLayoutCommon`](https://github.com/pytorch/glow/blob/master/include/glow/Graph/TensorLayout.h).
+Which includes the following virtual methods they can override:
```

Comment:
Could you add some wording about why backends may want to override such methods?

Reply:
Added OpenCL convolutions as an example for why a backend may want to do that.